README.md
# carrierwave-postgresql-table

[![Circle CI](https://circleci.com/gh/GUI/carrierwave-postgresql-table.svg?style=svg)](https://circleci.com/gh/GUI/carrierwave-postgresql-table)

A PostgreSQL storage adapter for [CarrierWave](https://github.com/carrierwaveuploader/carrierwave). Files are stored as PostgreSQL [large objects](https://www.postgresql.org/docs/current/static/largeobjects.html).

This gem is similar to [carrierwave-postgresql](https://github.com/diogob/carrierwave-postgresql), but differs in how it stores file metadata. This gem uses an extra table to store additional file metadata, allowing it to support multiple CarrierWave [versions](https://github.com/carrierwaveuploader/carrierwave#adding-versions) on a single uploader.

## Installation

Add this line to your application's Gemfile:

```ruby
gem "carrierwave-postgresql-table"
```

## Compatibility

carrierwave-postgresql-table is tested against the following versions of CarrierWave, Rails, and Ruby:

- CarrierWave 0.11, 1.0, 1.1, 1.2
- Rails 4.2, 5.0, 5.1, 5.2
- Ruby 2.2, 2.3, 2.4, 2.5

## Usage

Generate the table to store the file uploads:

```sh
$ rails generate carrierwave_postgresql_table:migration
$ rake db:migrate
```

Adjust your uploader to use the `postgresql_table` storage adapter:

```ruby
class AvatarUploader < CarrierWave::Uploader::Base
  storage :postgresql_table

  def store_dir
    "uploads/#{model.class.to_s.underscore}/#{mounted_as}/#{model.id}"
  end
end
```

An optional Rack application is available to provide an endpoint that will stream your files from PostgreSQL. To enable this endpoint, you can mount it in your `config/routes.rb` file:

```ruby
Rails.application.routes.draw do
  mount CarrierWave::PostgresqlTable::RackApp.new => "/uploads"
end
```

The Rack application can also be combined with other routing functionality. For example, to limit uploads to authenticated users when used with Devise:

```ruby
Rails.application.routes.draw do
  authenticate :user do
    mount CarrierWave::PostgresqlTable::RackApp.new => "/uploads"
  end
end
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at [https://github.com/GUI/carrierwave-postgresql-table](https://github.com/GUI/carrierwave-postgresql-table).

## Acknowledgments

Hat tip to the people behind [carrierwave-postgresql](https://github.com/diogob/carrierwave-postgresql) and [refile-postgres](https://github.com/krists/refile-postgres), which this code is based on.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# Change Log

## [1.1.0] - 2018-11-22

### Fixed
- Compatibility with Rails 5.1 and 5.2

### Changed
- Drop Ruby 1.9 compatibility.

## [1.0.2] - 2017-04-26

### Fixed
- Another fix for 404 handling in the Rack application. Thanks to [@bauersman](https://github.com/bauersman) for the fix.

## [1.0.1] - 2017-04-26

### Fixed
- Fix 404 handling in the Rack application. Thanks to [@bauersman](https://github.com/bauersman) for the fix.

## [1.0.0] - 2017-04-09

### Fixed
- Fix the Rack application when running with Rails under a relative URL root.
- Fix the Rack application to return 404 not found responses for unknown files.

### Changed
- Test against final, released version of CarrierWave 1.0.0.

## [0.1.3] - 2016-09-28

### Fixed
- Don't delete non-cached files when cleaning the cache.

## [0.1.2] - 2016-09-25

### Fixed
- Fix when connecting as a non-superuser account to PostgreSQL.

## [0.1.1] - 2016-09-25

### Fixed
- Fix `move_to` resulting in duplicate records.
- Fix `copy_to` leaving left over records in table.

## [0.1.0] - 2016-09-25

### Added
- Initial release.


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "carrierwave-postgresql-table"

TEST_ROOT = File.expand_path(__dir__)
Dir[File.join(TEST_ROOT, "support/**/*.rb")].sort.each { |f| require f }

require "minitest/autorun"


test/rack_app_test.rb
# frozen_string_literal: true

require "test_helper"
require "rack/test"

class RackAppTest < Minitest::Test
  include Rack::Test::Methods

  def setup
    DatabaseCleaner.start
  end

  def teardown
    DatabaseCleaner.clean
  end

  def app
    Rack::Lint.new(CarrierWave::PostgresqlTable::RackApp.new)
  end

  def test_returns_file
    user = User.create(:bio => File.new(File.join(TEST_ROOT, "fixtures/hello.txt")))
    file = CarrierWave::Storage::PostgresqlTable::CarrierWaveFile.first

    get "/uploads/user/bio/#{user.id}/hello.txt"
    assert_equal(200, last_response.status)
    assert_equal(file.updated_at.httpdate, last_response.headers["Last-Modified"])
    assert_equal("text/plain", last_response.headers["Content-Type"])
    assert_equal("inline", last_response.headers["Content-Disposition"])
    assert_equal("Hello, World.\n\n", last_response.body)
  end

  def test_optional_download
    user = User.create(:bio => File.new(File.join(TEST_ROOT, "fixtures/hello.txt")))

    get "/uploads/user/bio/#{user.id}/hello.txt?download=true"
    assert_equal(200, last_response.status)
    assert_equal("Hello, World.\n\n", last_response.body)
    assert_equal("attachment; filename=hello.txt", last_response.headers["Content-Disposition"])
  end

  def test_rails_relative_url_root
    Rails.application = TestRailsApp.new
    Rails.application.config.relative_url_root = "/foo/bar"
    user = User.create(:bio => File.new(File.join(TEST_ROOT, "fixtures/hello.txt")))

    get "/foo/bar/uploads/user/bio/#{user.id}/hello.txt"
    assert_equal(200, last_response.status)
    assert_equal("Hello, World.\n\n", last_response.body)
  ensure
    Rails.application.config.relative_url_root = nil
    Rails.application = nil
  end

  def test_rails_relative_url_root_trailing_slash
    Rails.application = TestRailsApp.new
    Rails.application.config.relative_url_root = "/foo/bar/"
    user = User.create(:bio => File.new(File.join(TEST_ROOT, "fixtures/hello.txt")))

    get "/foo/bar/uploads/user/bio/#{user.id}/hello.txt"
    assert_equal(200, last_response.status)
    assert_equal("Hello, World.\n\n", last_response.body)
  ensure
    Rails.application.config.relative_url_root = nil
    Rails.application = nil
  end

  def test_rails_relative_url_root_env
    ENV["RAILS_RELATIVE_URL_ROOT"] = "/foo/baz"
    Rails.application = TestRailsApp.new
    user = User.create(:bio => File.new(File.join(TEST_ROOT, "fixtures/hello.txt")))

    get "/foo/baz/uploads/user/bio/#{user.id}/hello.txt"
    assert_equal(200, last_response.status)
    assert_equal("Hello, World.\n\n", last_response.body)
  ensure
    ENV["RAILS_RELATIVE_URL_ROOT"] = nil
    Rails.application = nil
  end

  def test_not_found
    get "/uploads/foo/hello.txt"
    assert_equal(404, last_response.status)
    assert_equal("text/plain", last_response.headers["Content-Type"])
    assert_equal("Not Found", last_response.body)
  end
end


test/support/models.rb
# frozen_string_literal: true

class UserBioUploader < CarrierWave::Uploader::Base
  def store_dir
    "uploads/#{model.class.to_s.underscore}/#{mounted_as}/#{model.id}"
  end
end

class AnimalBioUploader < CarrierWave::Uploader::Base
  version :stripped do
    process :strip_text
  end

  def store_dir
    "uploads/#{model.class.to_s.underscore}/#{mounted_as}/#{model.id}"
  end

  private

  def strip_text
    cache_stored_file! unless cached?
    data = File.read(current_path).strip
    File.open(current_path, "wb") { |f| f.write(data) }
  end
end

class CatBioUploader < CarrierWave::Uploader::Base
  def store_dir
    "uploads/#{model.class.to_s.underscore}/#{mounted_as}/#{model.id}"
  end

  def move_to_store
    true
  end
end

class User < ActiveRecord::Base
  mount_uploader :bio, UserBioUploader

  validates :legacy_code, :allow_nil => true, :format => {
    :with => /\A[a-zA-Z]+\z/,
    :message => "only allows letters",
  }
end

class Animal < ActiveRecord::Base
  mount_uploader :bio, AnimalBioUploader
end

class Cat < ActiveRecord::Base
  self.table_name = "animals"
  mount_uploader :bio, CatBioUploader

  validates :legacy_code, :allow_nil => true, :format => {
    :with => /\A[a-zA-Z]+\z/,
    :message => "only allows letters",
  }
end


lib/carrierwave/storage/postgresql_table.rb
# frozen_string_literal: true

module CarrierWave
  module Storage
    class PostgresqlTable < Abstract
      def store!(file)
        if (uploader.move_to_store && file.kind_of?(CarrierWave::Storage::PostgresqlTable::File))
          file.move_to(uploader.store_path)
          file
        else
          f = CarrierWave::Storage::PostgresqlTable::File.new(uploader.store_path)
          f.store(file)
          f
        end
      end

      def retrieve!(identifier)
        CarrierWave::Storage::PostgresqlTable::File.new(uploader.store_path(identifier))
      end

      def cache!(new_file)
        f = CarrierWave::Storage::PostgresqlTable::File.new(uploader.cache_path)
        f.store(new_file)
        f
      end

      def retrieve_from_cache!(identifier)
        CarrierWave::Storage::PostgresqlTable::File.new(uploader.cache_path(identifier))
      end

      def delete_dir!(path)
        # This is only supposed to delete *empty* directories, which don't
        # exist in our database.
      end

      def clean_cache!(seconds)
        time = Time.now - seconds.seconds
        CarrierWaveFile.delete_all_files("path LIKE #{CarrierWaveFile.connection.quote(::File.join(uploader.cache_dir, "%"))} AND updated_at < #{CarrierWaveFile.connection.quote(time)}")
      end

      class CarrierWaveFile < ::ActiveRecord::Base
        self.table_name = "carrierwave_files"

        def self.delete_all_files(conditions)
          self.transaction do
            self.connection.execute("SELECT lo_unlink(pg_largeobject_oid) FROM (SELECT DISTINCT pg_largeobject_oid FROM #{self.table_name} WHERE #{conditions}) AS oids")
            self.where(conditions).delete_all
          end
        end
      end

      class File
        READ_CHUNK_SIZE = 16_384
        STREAM_CHUNK_SIZE = 16_384

        attr_reader :path

        def initialize(path)
          @path = path
          @record = CarrierWaveFile.find_or_initialize_by(:path => path)
          @read_pos = 0
        end

        def read(length = nil, buffer = nil)
          data = nil
          CarrierWaveFile.transaction do
            raw_connection = CarrierWaveFile.connection.raw_connection

            begin
              lo = raw_connection.lo_open(@record.pg_largeobject_oid)
              if length
                raw_connection.lo_lseek(lo, @read_pos, PG::SEEK_SET)
                data = raw_connection.lo_read(lo, length)
                @read_pos = raw_connection.lo_tell(lo)
              else
                data = raw_connection.lo_read(lo, self.size)
              end
            ensure
              raw_connection.lo_close(lo) if lo
            end
          end

          if (buffer && data)
            buffer.replace(data)
          end

          data
        end

        def to_tempfile
          Tempfile.new(:binmode => true).tap do |tempfile|
            IO.copy_stream(self, tempfile)
            self.rewind
            tempfile.rewind
            tempfile.fsync
          end
        end

        def filename
          ::File.basename(@path)
        end

        def last_modified
          @record.updated_at
        end

        def size
          @size = @record.size || fetch_size
        end

        def eof?
          @read_pos == self.size
        end

        def rewind
          @read_pos = 0
        end

        def content_type
          @record.content_type
        end

        def content_type=(new_content_type)
          @record.update_attribute(:content_type, new_content_type)
        end

        def url(_options = {})
          ::File.join("/", @path)
        end

        def store(new_file)
          CarrierWaveFile.transaction do
            connection = CarrierWaveFile.connection
            raw_connection = connection.raw_connection
            oid = nil
            if new_file.kind_of?(CarrierWave::Storage::PostgresqlTable::File)
              file = new_file
            else
              file = new_file.to_file
            end

            begin
              oid = @record.pg_largeobject_oid || raw_connection.lo_creat
              handle = raw_connection.lo_open(oid, PG::INV_WRITE)
              raw_connection.lo_truncate(handle, 0)
              buffer = String.new # rubocop:disable Style/EmptyLiteral
              until file.eof?
                file.read(READ_CHUNK_SIZE, buffer)
                raw_connection.lo_write(handle, buffer)
              end
              file.rewind
            ensure
              raw_connection.lo_close(handle)
            end

            begin
              old_oid = @record.pg_largeobject_oid
              @record.pg_largeobject_oid = oid
              @record.size = new_file.size
              @record.content_type = new_file.content_type
              @record.save

              # Cleanup old, unused largeobject OIDs if we're updating the
              # record with a new OID reference.
              if (old_oid && old_oid != oid)
                old_references = connection.select_value("SELECT COUNT(*) FROM #{CarrierWaveFile.table_name} WHERE pg_largeobject_oid = #{CarrierWaveFile.connection.quote(old_oid)}").to_i
                if (old_references == 0)
                  raw_connection.lo_unlink(old_oid)
                end
              end
            rescue ::ActiveRecord::RecordNotUnique
              @record = CarrierWaveFile.find_or_initialize_by(:path => @path)
              retry
            end
          end
        end

        def exists?
          @record && @record.persisted?
        end

        def move_to(new_path)
          CarrierWaveFile.transaction do
            # Remove any existing files at the current path.
            CarrierWaveFile.delete_all_files("path = #{CarrierWaveFile.connection.quote(new_path)} AND id != #{CarrierWaveFile.connection.quote(@record.id)}")

            # Change the current record's path to the new path.
            @record.update_attribute(:path, new_path)
          end
        end

        def delete
          CarrierWaveFile.delete_all_files("id = #{CarrierWaveFile.connection.quote(@record.id)}")
        end

        private

        def fetch_size
          size = nil
          CarrierWaveFile.transaction do
            raw_connection = CarrierWaveFile.connection.raw_connection
            lo = raw_connection.lo_open(@record.pg_largeobject_oid)
            size = raw_connection.lo_lseek(lo, 0, PG::SEEK_END)
          end

          size
        end
      end
    end
  end
end


lib/carrierwave/postgresql_table/rack_app.rb
# frozen_string_literal: true

module CarrierWave
  module PostgresqlTable
    class RackApp
      READ_CHUNK_SIZE = 16_384

      def call(env)
        request = Rack::Request.new(env)

        strip_prefix = "/"
        if (defined?(Rails) && Rails.application && Rails.application.config.relative_url_root)
          strip_prefix = File.join(Rails.application.config.relative_url_root, "/")
        end
        path = request.path.sub(/^#{Regexp.escape(strip_prefix)}/, "")
        file = CarrierWave::Storage::PostgresqlTable::File.new(path)

        unless file.exists?
          return [404, { "Content-Type" => "text/plain" }, ["Not Found"]]
        end

        headers = {
          "Last-Modified" => file.last_modified.httpdate,
          "Content-Type" => file.content_type,
          "Content-Disposition" => "inline",
        }

        if (request.params["download"] == "true")
          headers["Content-Disposition"] = "attachment; filename=#{file.filename}"
        end

        body = Enumerator.new do |response_body|
          while (chunk = file.read(READ_CHUNK_SIZE))
            response_body << chunk
          end
        end

        [200, headers, body]
      end
    end
  end
end
